### DISCUSSION TITLE
HowTo: Upgrade manual Installation

### DISCUSSION THREAD

**Discussion by manuelkamp** [teslamate-org/teslamate#2667] [Q&A] [answered]
Hi, Is there any full documented HowTo for manual installations (on Debian)? The existing Manual fails already on the first command showed there:
`git pull`
`You are not currently on a branch.
Please specify which branch you want to merge with.`

I have no idea what to put there, since I never used git (oldschool svn User here).

**Comment by ptar** [comment]
Would you like to install Teslamate or to get the github repository?

For the installation, I would recommend this [documentation of the Teslamate Docker install](https://docs.teslamate.org/docs/installation/docker). If you save all the needed tokens and passwords in the right place (e.g. Keepass or alike) and don't leave blanks in the docker-compose.yml while inserting them, it should work almost instantly.
When it comes to Grafana: Add the Grafana-URL in the Teslamate-Settings to see the Dashboards-menu and really remember the admin password (->Keepass!) you'll have to choose after the first login!

For the github repository: https://docs.github.com seems to be a good starting point...

**Reply by manuelkamp** [reply]
Neither. As I stated in the title, I am searching for a HowTo for Upgrades for manual installations. So I have a manual installation and want to perform an upgrade, but the first command on the Teslamate Manual on Upgrades gives me the error I posted.

**Reply by ptar** [reply]
Ups, sorry, I missed that, because I've never walked that path. For me, the encapsulation with a docker installation always seemed to be the best way to use and maintain Teslamate.
That said, my only answer to your question would be the recommendation to do a backup and restore it in a docker installation. For Raspbian, I used [this easy guide](https://dev.to/rohansawant/installing-docker-and-docker-compose-on-the-raspberry-pi-in-5-simple-steps-3mgl) to install the Docker environment. For Debian it's not just adding the repository to the sources and running an update/upgrade, see [here](https://docs.docker.com/engine/install/debian/#installation-methods)

Now I pass the question on to someone who could help you with the manual installation...

**Reply by manuelkamp** [reply]
No problem. Docker is not an option for me, since I have Teslamate running in a LXC on my HA Proxmox Cluster - and no Docker Containers at all in my service.

**Comment by cwanja** [comment]
Have you tried to specify a branch? `git branch`

**Reply by manuelkamp** [reply]
I tried it now, but it does not work it seems (sorry if I missed something, but I am not used into git and have no clue about it, since i stick to svn ever since)

`root@srv-tesla:/usr/src/teslamate# git branch
* (HEAD detached at v1.26.1)
  master
root@srv-tesla:/usr/src/teslamate# git pull
You are not currently on a branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>`

**Reply by cwanja** [reply]
> I tried it now, but it does not work it seems (sorry if I missed something, but I am not used into git and have no clue about it, since i stick to svn ever since)
> 
> 
> 
> `root@srv-tesla:/usr/src/teslamate# git branch
> 
> * (HEAD detached at v1.26.1)
> 
>   master
> 
> root@srv-tesla:/usr/src/teslamate# git pull
> 
> You are not currently on a branch.
> 
> Please specify which branch you want to merge with.
> 
> See git-pull(1) for details.
> 
> 
> 
>     git pull <remote> <branch>`

Have you also looked at the manual Debian install? https://docs.teslamate.org/docs/installation/debian

One step seems to be to clone the git repo. Pull (I would think) would be if you are going to modify it. Seems like you just want to execute the application.

**Reply by manuelkamp** [reply]
I am not sure what you mean. I am not looking for a new installation, since I already have it installed manually several months ago (followed all steps in the docs you linked). So I already have it cloned in /usr/src/teslamate (this still is there from months ago). Now what I did is to go into this directory and followed strictly the Upgrade Docs, but they fail on the first step...

**Reply by cwanja** [reply]
> I am not sure what you mean. I am not looking for a new installation, since I already have it installed manually several months ago (followed all steps in the docs you linked). So I already have it cloned in /usr/src/teslamate (this still is there from months ago). Now what I did is to go into this directory and followed strictly the Upgrade Docs, but they fail on the first step...

I am certainly no Git expert. But it seems like you need to define the branch you want to pull from either in the commands or in a configuration file. I know the documents do not reference that, but manual installs could be different by system. My suggestion would be to look at `git branch` or `git pull` help pages to define a branch you want to pull from. 

Or take a backup and start completely over. Most invasive route.

**Reply by mnadvornik** [reply]
@manuelkamp you can either remove the git repository and clone it again as described [here](https://docs.teslamate.org/docs/installation/debian/#clone-teslamate-git-repository) or you can fetch the current tags and check out the latest version using these commands:
```
cd /usr/src/teslamate
git fetch --all
git checkout $(git describe --tags `git rev-list --tags --max-count=1`)
```
Then just recompile as described [here](https://docs.teslamate.org/docs/installation/debian/#compile-elixir-project)

You can also import the current Grafana dashboards again in case they have changed since the version you are using.

**Reply by manuelkamp** [reply]
thanks mate, that worked perfect. Now I am on 1.26.1

**Comment by manuelkamp** [comment, marked as answer]
To answer and close this thread, look at the wonderful answer from @mnadvornik https://github.com/adriankumpf/teslamate/discussions/2667#discussioncomment-2850274
